### ISSUE TITLE
Very high memory usage due to caching preprocessed output

### ISSUE THREAD

**Issue by avudnez** [fastbuild/fastbuild#912]
Hi,

We use distribution, cache, but **not** PCHs.
When building a lot of TUs, FBuild.exe memory consumption skyrockets and it's not uncommon it reaches 15+GB. At which point it is sabotaging itself because the system doesn't have enough memory left to build correctly.

This is due to, if I understand correctly, fbuild keeping the (compressed) preprocessed output in memory.
Example callstack of such allocations, monitored with [UMDH](https://docs.microsoft.com/en-us/windows-hardware/drivers/debugger/using-umdh-to-find-a-user-mode-memory-leak):
```
	ntdll!RtlAllocateHeap+AFD
	fbuild!_malloc_base+44 (minkernel\crts\ucrt\src\appcrt\heap\malloc_base.cpp, 34)
	fbuild!_aligned_malloc+49 (minkernel\crts\ucrt\src\appcrt\heap\align.cpp, 555)
	fbuild!Compressor::Compress+8D (fastbuild\Code\Tools\FBuild\FBuildCore\Helpers\Compressor.cpp, 99)
	fbuild!ObjectNode::DoBuildWithPreProcessor+5DD (fastbuild\Code\Tools\FBuild\FBuildCore\Graph\ObjectNode.cpp, 519)
	fbuild!ObjectNode::DoBuild+224 (fastbuild\Code\Tools\FBuild\FBuildCore\Graph\ObjectNode.cpp, 245)
	fbuild!JobQueue::DoBuild+144 (fastbuild\Code\Tools\FBuild\FBuildCore\WorkerPool\JobQueue.cpp, 796)
	fbuild!WorkerThread::Update+152 (fastbuild\Code\Tools\FBuild\FBuildCore\WorkerPool\WorkerThread.cpp, 214)
	fbuild!WorkerThread::Main+338 (fastbuild\Code\Tools\FBuild\FBuildCore\WorkerPool\WorkerThread.cpp, 142)
	fbuild!WorkerThread::ThreadWrapperFunc+2C (fastbuild\Code\Tools\FBuild\FBuildCore\WorkerPool\WorkerThread.cpp, 128)
```

It seems the output is also [written to disk](https://github.com/fastbuild/fastbuild/blob/b763082a4752b87122999f6d569b76c9f897eea1/Code/Tools/FBuild/FBuildCore/Graph/ObjectNode.cpp#L2095). With the _massive_ TUs of Unreal Engine, preprocessed output easily reaches 35MB or more. Over hundreds or thousands of TUs, this means a lot of memory.
What's the rationale for keeping it in memory ? Could we free it instead once it's written to disk ?

Note: using lightcache would help but we can't really for now because of #893 and #894.

**Comment by ffulin** [comment]
FASTBuild runs the preprocessor in various situations, most notably in your case to support distributed compilation.
The callstack shown here is storing the preprocessed data in order to be able to send it remotely.
The data is discarded when that job is complete (locally or remotely) and the memory is freed.

This data is also subject to a memory limit which is 2GiB by default (can be overridden in the bff Settings()) in order to prevent unbounded growth.

The actual memory used could conceivably be higher than the limit during compression by (numCPUCoresDoingCompression * sizeof(preprocessed data))

If you are seeing 15GiB peak it seems there is something unusual going on. Even with 35MiB preprocessed object files to see a 15GiB footprint would require about 200 cpu cores all simultaneously completing a job and compressing the data at the same time. it seems unlikely that this would happen.

I can think of a few possible explanations:
a) DistributableJobMemoryLimitMiB has been increased, possibly by some 3rd party bff generator used for Unreal integration.
This can be checked by examining the bff file.
b) While some memory certainly comes from the stack you mentioned, perhaps there is an unexpectedly large memory usage somewhere else
c) There is a bug and in some situation this memory is not being managed properly (a fault in the memory use tracking or a leak for example)

If you run the build with the -profile command line you can see the memory footprint of FASTBuild (total system memory use as reported by the OS) as well as what it thinks the local memory use is for objects pending distributed compilation.

*> Note: using lightcache would help but we can't really for now because of https://github.com/fastbuild/fastbuild/issues/893 and https://github.com/fastbuild/fastbuild/issues/894.*

The LightCache should not change this behavior as this preprocessed data is used for distributed compilation as noted above.

**Comment by avudnez** [comment]
Hi, I realized I didn't update this issue yet I found my problem, thanks for the ping.
The thing was that an unrelated (questionable) patch added a code path where the preprocessed output waiting to be distributed was kept in memory for future use (and not compiled locally), which is not the case in vanilla FBuild.
With thousands of enormous TUs it was eating RAM very quickly.
Now we compress it while it's not used.

Sorry for the confusion, I think we can close this issue.
